File: android/app/src/main/java/com/conxius/wallet/NativeCrypto.java
package com.conxius.wallet;

import android.util.Base64;
import org.json.JSONArray;
import org.json.JSONObject;
import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;

public class NativeCrypto {
    private static final int GCM_TAG_BITS = 128;

    public static byte[] decryptVault(String vaultJson, String pin) throws Exception {
        JSONObject envelope = new JSONObject(vaultJson);
        int v = envelope.getInt("v");
        if (v != 1) throw new IllegalArgumentException("Unknown vault version");

        JSONArray saltJson = envelope.getJSONArray("salt");
        JSONArray ivJson = envelope.getJSONArray("iv");
        JSONArray dataJson = envelope.getJSONArray("data");

        byte[] salt = new byte[saltJson.length()];
        for(int i=0; i<saltJson.length(); i++) salt[i] = (byte)saltJson.getInt(i);

        byte[] iv = new byte[ivJson.length()];
        for(int i=0; i<ivJson.length(); i++) iv[i] = (byte)ivJson.getInt(i);

        byte[] data = new byte[dataJson.length()];
        for(int i=0; i<dataJson.length(); i++) data[i] = (byte)dataJson.getInt(i);

        // Derive Key: PBKDF2WithHmacSHA256, 200000 iterations
        SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");
        PBEKeySpec spec = new PBEKeySpec(pin.toCharArray(), salt, 200000, 256);
        SecretKey tmp = factory.generateSecret(spec);
        SecretKeySpec secret = new SecretKeySpec(tmp.getEncoded(), "AES");

        // Decrypt: AES/GCM/NoPadding
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        cipher.init(Cipher.DECRYPT_MODE, secret, new GCMParameterSpec(GCM_TAG_BITS, iv));
        
        return cipher.doFinal(data);
    }
}

File: android/app/src/main/kotlin/com/conxius/wallet/NativeCrypto.kt
package com.conxius.wallet

import org.json.JSONObject
import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec
import java.util.Arrays

object NativeCrypto {
    private const val GCM_TAG_BITS = 128

    fun decryptVault(vaultJson: String, pin: String): ByteArray {
        val envelope = JSONObject(vaultJson)
        val v = envelope.getInt("v")
        if (v != 1) throw IllegalArgumentException("Unknown vault version")

        val saltJson = envelope.getJSONArray("salt")
        val ivJson = envelope.getJSONArray("iv")
        val dataJson = envelope.getJSONArray("data")

        val salt = ByteArray(saltJson.length()) { saltJson.getInt(it).toByte() }
        val iv = ByteArray(ivJson.length()) { ivJson.getInt(it).toByte() }
        val data = ByteArray(dataJson.length()) { dataJson.getInt(it).toByte() }

        // Derive Key: PBKDF2WithHmacSHA256, 200000 iterations
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        val pinChars = pin.toCharArray()
        val spec = PBEKeySpec(pinChars, salt, 200000, 256)
        val tmp = factory.generateSecret(spec)

        // Zero-out PIN array immediately
        Arrays.fill(pinChars, '\u0000')
        spec.clearPassword()

        val encodedKey = tmp.encoded
        val secret = SecretKeySpec(encodedKey, "AES")
        Arrays.fill(encodedKey, 0.toByte())

        // Decrypt: AES/GCM/NoPadding
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, secret, GCMParameterSpec(GCM_TAG_BITS, iv))

        return cipher.doFinal(data)
    }
}
